README.md
Image Scan! with 16 bits scanning capability for Ubuntu 20.04
=============================================================

The ImageScan! backend for Linux does not support 16 bits scanning.
Henrik Andersson fixed that a while ago in his `iscan` fork:

https://github.com/hean01/iscan

However, the Linux environment evolved since then and it does
not compile on my current Ubuntu 20.04 environment.

Thanks to

- https://bbs.archlinux.org/viewtopic.php?id=244017
- http://www.slackware.com/~alien/slackbuilds/iscan/build/

I was able to fix a few of the issues. I used the debugger to fix
the other ones I found.

As the title says, these instructions are for Ubuntu, more precisely
Ubuntu 20.04. They may work as is or with modifications in other
environments.

Please also read the original [README](README) file from Epson.

Installing
==========

The installation method that worked for me was to install this package
over the Epson drivers. So first, go on the
[Epson website](https://download.ebz.epson.net/dsc/search/01/search/searchModule)
and download the drivers for your scanner. This fork is based on
iscan 2.30.4-2 so you would need to install iscan 2.30.4-2 from Epson.

Test your installation. Make sure you are able to scan an 8 bit image
with `iscan` or `xsane`. Check that the software displays the scan
options that you would expect from your scanner (transparency unit, etc.)
If the expected options are not there, your scanner was not recognized.
In this case, seek help on the Internet.
[This page](https://nzeid.net/linux-sane-epson-perfection-v600-photo-scanner-setup)
may help you for the V600.

Once you can scan with the Epson drivers, you need to compile this package
and install it over Epson `iscan`.

Compiling
=========

Make a directory called `iscan-16` and `cd` into it. Then clone this repository:

```
mkdir iscan-16
cd iscan-16
git clone https://github.com/patrickfournier/iscan.git
```

Then compile the package:

```
cd iscan
git checkout master
ln -s libesmod-x86_64.c2.so non-free/libesmod-x86_64.so
dpkg-buildpackage -rfakeroot --no-sign
```

If there are any missing dependency reported by `dpkg-buildpackage`, install them
and run `dpkg-buildpackage` again.

Installing
==========

At this point, you should have a file named `iscan_2.30.4-2_amd64.deb` in the `iscan-16`
directory. Simply install it over the existing `iscan` package (do not remove Epson
package before installing this one, as it may contain a plugin specific to your scanner):

`sudo dpkg --install iscan_2.30.4-2_amd64.deb`

Extra notes
===========

These are notes for myself.

IScan sources
-------------

http://support.epson.net/linux/src/scanner/iscan/

Dependencies
------------

cdbs
debhelper (>= 9)
libsane-dev
libltdl7-dev | libltdl3-dev
libgtk2.0-dev
libpng-dev
libjpeg-dev
libgimp2.0-dev
libtiff5-dev
libxml2-dev
libusb-1.0-0-dev

Compiling manually
-----------------

```
make distclean

LDFLAGS="-ldl -lpng" \
./configure \
  --prefix=/usr \
  --sbindir=/usr/bin \
  --libdir=/usr/lib${LIBDIRSUFFIX} \
  --mandir=/usr/man \
  --docdir=/usr/doc/$PRGNAM-$VERSION \
  --localstatedir=/var \
  --sysconfdir=/etc \
  --enable-dependency-reduction \
  --enable-frontend \
  --enable-jpeg \
  --enable-tiff \
  --enable-png \
  --enable-gimp \
  --program-prefix= \
  --program-suffix=

make
```

Manual install
--------------

- Install epsonscan2
- Install iscan
- `make DESTDIR=<some-dir-abs-path> install-strip`

```
cd /usr/bin
sudo cp <some-dir-abs-path>/usr/bin/iscan .

cd /usr/lib/sane
sudo cp <some-dir-abs-path>/usr/lib/sane/libsane-epkowa.la .
sudo cp <some-dir-abs-path>/usr/lib/sane/libsane-epkowa.so.1.0.15 .

cd /usr/sbin
sudo cp <some-dir-abs-path>/usr/sbin/iscan-registry .
```

Debugging Epkowa
----------------

`export SANE_DEBUG_EPKOWA=<VALUE>`

with <VALUE> one of

```
FATAL
MAJOR
MINOR
INFO
CALL
DATA
CMD
HEX
IMG
```


debian/changelog
iscan (2.30.4-2) unstable; urgency=low

  * latest "upstream"
    See the NEWS and/or ChangeLog files for details.

 -- SEIKO EPSON Corporation <linux-printer@epson.jp>  Thu, 06 Jan 2022 14:56:01 -0500


lib/pngstream.cc
//  pngstream.cc -- image streams producing PNG files
//  Copyright (C) 2008, 2009, 2016  SEIKO EPSON CORPORATION
//
//  This file is part of the 'iscan' program.
//
//  The 'iscan' program is free-ish software.
//  You can redistribute it and/or modify it under the terms of the GNU
//  General Public License as published by the Free Software Foundation;
//  either version 2 of the License or at your option any later version.
//
//  This program is distributed in the hope that it will be useful, but
//  WITHOUT ANY WARRANTY;  without even the implied warranty of FITNESS
//  FOR A PARTICULAR PURPOSE or MERCHANTABILITY.
//  See the GNU General Public License for more details.
//
//  You should have received a verbatim copy of the GNU General Public
//  License along with this program; if not, write to:
//
//      Free Software Foundation, Inc.
//      59 Temple Place, Suite 330
//      Boston, MA  02111-1307  USA
//
//  As a special exception, the copyright holders give permission
//  to link the code of this program with the esmod library and
//  distribute linked combinations including the two.  You must obey
//  the GNU General Public License in all respects for all of the
//  code used other then esmod.


#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "pngstream.hh"

#include <iostream>

namespace iscan
{
  pngstream::pngstream (FILE *fp, const string& name)
    : _stream (fp), _header (false), _footer (false)
  {
    if (!_stream) throw std::invalid_argument ("invalid file handle");
#if HAVE_PNG_H
    init ();
#endif
  }

  pngstream::~pngstream (void)
  {
    try
      {
        flush ();
      }
    catch (const std::exception& oops)
      {
        std::cerr << oops.what ();
      }

#if HAVE_PNG_H
    lib->destroy_write_struct (&_png, &_info);
#endif
  }

  basic_imgstream&
  pngstream::write (const byte_type *line, size_type n)
  {
    if (!line || 0 == n) return *this;
#if HAVE_PNG_H
    if (!_header)
      {
        write_header ();
      }
    set_error_handler (_png, _info);
    lib->write_row (_png, (png_byte *) line);
#endif
    return *this;
  }

  basic_imgstream&
  pngstream::flush (void)
  {
#if HAVE_PNG_H
    set_error_handler (_png, _info);

    /* when not interlacing (ie, only one pass), number of rows is image height:  _v_sz */
#if PNG_LIBPNG_VER > 10499
    if (!_footer && _v_sz == lib->get_current_row_number(_png))
#else
    if (_header && !_footer && _png->num_rows == _png->flush_rows)
#endif
      {
        lib->write_end (_png, _info);
        _footer = true;
      }
#endif
    fflush (_stream);

    return *this;
  }

  bool
  pngstream::is_usable (void)
  {
    if (lib)
      {
        return lib->is_usable;
      }

    lib = new (std::nothrow) png_lib_handle ();
    if (!lib)
      {
        return false;
      }

    lib->is_usable = false;
    lib->message   = string ();
    lib->lib       = NULL;
#if HAVE_PNG_H
    try
      {
        // Some PNG libraries are *not* linked with libz which leads
        // to a "file not found" error when calling dlopen.  We just
        // dlopen the libz library explicitly to work around this.
        basic_imgstream::dlopen ("libz");
        basic_imgstream::dlopen ("libpng", validate);
      }
    catch (std::runtime_error& e)
      {
        lib->message = e.what ();
        return lib->is_usable;
      }
#endif /* HAVE_PNG_H */

    return lib->is_usable;
  }

#if HAVE_PNG_H
#define funcsym(name)                                   \
  lib->name                                             \
    = ((pngstream::png_lib_handle::name##_f)            \
       basic_imgstream::dlsym (lib->lib, "png_"#name));
#endif

  bool
  pngstream::validate (lt_dlhandle h)
  {
    if (!h) return false;

#if HAVE_PNG_H
    lib->lib = h;

    funcsym (access_version_number);
    funcsym (create_write_struct);
    funcsym (create_info_struct);
    funcsym (destroy_write_struct);
    funcsym (init_io);
    funcsym (set_IHDR);
    funcsym (set_pHYs);
    funcsym (set_invert_mono);
    funcsym (write_info);
    funcsym (write_row);
    funcsym (write_flush);
    funcsym (write_end);
#if PNG_LIBPNG_VER > 10499
    funcsym (get_current_row_number);
#endif

    if (lib->access_version_number
        && lib->create_write_struct
        && lib->create_info_struct
        && lib->destroy_write_struct
        && lib->init_io
        && lib->set_IHDR
        && lib->set_pHYs
        && lib->set_invert_mono
#if PNG_LIBPNG_VER > 10499
        && lib->get_current_row_number
#endif
        && lib->write_info
        && lib->write_row
        && lib->write_flush
        && lib->write_end)
      {
        lib->is_usable = (PNG_LIBPNG_VER <= lib->access_version_number ());
      }
#endif /* HAVE_PNG_H */

    return lib->is_usable;
  }

#if HAVE_PNG_H
#undef funcsym
#endif

  basic_imgstream&
  pngstream::write_header (void)
  {
    check_consistency ();

#if HAVE_PNG_H
    set_error_handler (_png, _info);

    if (mono == _cspc)
      {
        _bits = 1;
        lib->set_invert_mono (_png);
      }

    lib->init_io (_png, _stream);

    lib->set_IHDR (_png, _info, _h_sz, _v_sz, _bits,
                   (RGB == _cspc
                    ? PNG_COLOR_TYPE_RGB : PNG_COLOR_TYPE_GRAY),
                   PNG_INTERLACE_NONE,
                   PNG_COMPRESSION_TYPE_DEFAULT,
                   PNG_FILTER_TYPE_DEFAULT);

    size_type hres = size_type (100 / 2.54 * _hres + 0.5);
    size_type vres = size_type (100 / 2.54 * _vres + 0.5);
    lib->set_pHYs (_png, _info, hres, vres, PNG_RESOLUTION_METER);

    lib->write_info (_png, _info);
#endif /* HAVE_PNG_H */

    _header = true;
    return *this;
  }

  void
  pngstream::check_consistency (void) const
  {
    if (!(mono == _cspc || grey == _cspc || RGB == _cspc))
      {
        throw std::logic_error ("unsupported colour space");
      }
  }

  void
  pngstream::init (void)
  {
    if (!is_usable ())
      {
        throw std::runtime_error (lib->message);
      }

#if HAVE_PNG_H
    _png  = lib->create_write_struct (version_string,
                                      NULL, NULL, NULL);
    _info = NULL;
    if (!_png) throw std::bad_alloc ();

    _info = lib->create_info_struct (_png);
    if (!_info)
      {
        lib->destroy_write_struct (&_png, NULL);
        throw std::bad_alloc ();
      }
#endif /* HAVE_PNG_H */
  }

  pngstream::png_lib_handle *pngstream::lib = NULL;

#if HAVE_PNG_H
  png_const_charp pngstream::version_string = PNG_LIBPNG_VER_STRING;
#endif

} // namespace iscan


backend/channel-scsi.c
/*  channel_scsi.c -- SCSI device communication channel
 *  Copyright (C) 2008, 2009, 2013  SEIKO EPSON CORPORATION
 *
 *  License: GPLv2+|iscan
 *  Authors: AVASYS CORPORATION
 *
 *  This file is part of the SANE backend distributed with Image Scan!
 *
 *  Image Scan!'s SANE backend is free software.
 *  You can redistribute it and/or modify it under the terms of the GNU
 *  General Public License as published by the Free Software Foundation;
 *  either version 2 of the License or at your option any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *  You ought to have received a copy of the GNU General Public License
 *  along with this package.  If not, see <http://www.gnu.org/licenses/>.
 *
 *
 *  Linking Image Scan!'s SANE backend statically or dynamically with
 *  other modules is making a combined work based on this SANE backend.
 *  Thus, the terms and conditions of the GNU General Public License
 *  cover the whole combination.
 *
 *  As a special exception, the copyright holders of Image Scan!'s SANE
 *  backend give you permission to link Image Scan!'s SANE backend with
 *  SANE frontends that communicate with Image Scan!'s SANE backend
 *  solely through the SANE Application Programming Interface,
 *  regardless of the license terms of these SANE frontends, and to
 *  copy and distribute the resulting combined work under terms of your
 *  choice, provided that every copy of the combined work is
 *  accompanied by a complete copy of the source code of Image Scan!'s
 *  SANE backend (the version of Image Scan!'s SANE backend used to
 *  produce the combined work), being distributed under the terms of
 *  the GNU General Public License plus this exception.  An independent
 *  module is a module which is not derived from or based on Image
 *  Scan!'s SANE backend.
 *
 *  As a special exception, the copyright holders of Image Scan!'s SANE
 *  backend give you permission to link Image Scan!'s SANE backend with
 *  independent modules that communicate with Image Scan!'s SANE
 *  backend solely through the "Interpreter" interface, regardless of
 *  the license terms of these independent modules, and to copy and
 *  distribute the resulting combined work under terms of your choice,
 *  provided that every copy of the combined work is accompanied by a
 *  complete copy of the source code of Image Scan!'s SANE backend (the
 *  version of Image Scan!'s SANE backend used to produce the combined
 *  work), being distributed under the terms of the GNU General Public
 *  License plus this exception.  An independent module is a module
 *  which is not derived from or based on Image Scan!'s SANE backend.
 *
 *  Note that people who make modified versions of Image Scan!'s SANE
 *  backend are not obligated to grant special exceptions for their
 *  modified versions; it is their choice whether to do so.  The GNU
 *  General Public License gives permission to release a modified
 *  version without this exception; this exception also makes it
 *  possible to release a modified version which carries forward this
 *  exception.
 */


/*! \file
    \brief  Implements a SCSI communication channel.
 */


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#define USE_PROTECTED_CHANNEL_API
#include "channel.h"

#include <string.h>

#include "utils.h"

/* Deprecated includes */
#include "epkowa_scsi.h"


static void channel_scsi_open  (channel *, SANE_Status *);
static void channel_scsi_close (channel *, SANE_Status *);

static ssize_t channel_scsi_send (channel *, const void *,
                                  size_t, SANE_Status *);
static ssize_t channel_scsi_recv (channel *, void *,
                                  size_t, SANE_Status *);

static void channel_scsi_set_max_request_size (channel *, size_t);


channel *
channel_scsi_ctor (channel *self, const char *dev_name, SANE_Status *status)
{
  size_t name_len = 0;

  require (self && dev_name);
  require (0 == strncmp_c (dev_name, "scsi:", strlen ("scsi:")));

  dev_name += strlen ("scsi:");
  name_len  = strlen (dev_name) + 1;

  self->name = t_malloc (name_len, char);
  if (!self->name)
    {
      if (status) *status = SANE_STATUS_NO_MEM;
      return self->dtor (self);
    }
  strcpy (self->name, dev_name);

  self->open  = channel_scsi_open;
  self->close = channel_scsi_close;

  self->send = channel_scsi_send;
  self->recv = channel_scsi_recv;

  self->set_max_request_size = channel_scsi_set_max_request_size;

  self->max_size = sanei_scsi_max_request_size;

  return self;
}

static ssize_t
channel_scsi_send (channel *self, const void *buffer,
                   size_t size, SANE_Status *status)
{
  return sanei_epson_scsi_write (self->fd, buffer, size, status);
}

static ssize_t
channel_scsi_recv (channel *self, void *buffer,
                   size_t size, SANE_Status *status)
{
  return sanei_epson_scsi_read (self->fd, buffer, size, status);
}

static void
channel_scsi_open (channel *self, SANE_Status *status)
{
  SANE_Status s = sanei_scsi_open (self->name, &self->fd,
                                   sanei_epson_scsi_sense_handler, NULL);
  if (SANE_STATUS_GOOD != s)
    {
      err_fatal ("can not open %s (%s)", self->name, sane_strstatus (s));
    }
  if (status) *status = s;
}

static void
channel_scsi_close (channel *self, SANE_Status *status)
{
  if (self->fd >= 0)
    {
      sanei_scsi_close (self->fd);
      self->fd = -1;
    }

  if (status) *status = SANE_STATUS_GOOD;
}

static void
channel_scsi_set_max_request_size (channel *self, size_t size)
{
  require (self);

  self->max_size =  (size < sanei_scsi_max_request_size)
                   ? size : sanei_scsi_max_request_size;
}


lib/jpegstream.cc
//  jpegstream.cc -- image streams producing JPEG files
//  Copyright (C) 2008, 2009  SEIKO EPSON CORPORATION
//
//  This file is part of the 'iscan' program.
//
//  The 'iscan' program is free-ish software.
//  You can redistribute it and/or modify it under the terms of the GNU
//  General Public License as published by the Free Software Foundation;
//  either version 2 of the License or at your option any later version.
//
//  This program is distributed in the hope that it will be useful, but
//  WITHOUT ANY WARRANTY;  without even the implied warranty of FITNESS
//  FOR A PARTICULAR PURPOSE or MERCHANTABILITY.
//  See the GNU General Public License for more details.
//
//  You should have received a verbatim copy of the GNU General Public
//  License along with this program; if not, write to:
//
//      Free Software Foundation, Inc.
//      59 Temple Place, Suite 330
//      Boston, MA  02111-1307  USA
//
//  As a special exception, the copyright holders give permission
//  to link the code of this program with the esmod library and
//  distribute linked combinations including the two.  You must obey
//  the GNU General Public License in all respects for all of the
//  code used other then esmod.


#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "jpegstream.hh"

#include <cstdlib>
#include <ios>

namespace iscan
{
  jpegstream::jpegstream (FILE *fp, const string& name)
    : _stream (fp), _header (false), _scanline (NULL)
  {
    if (!_stream) throw std::invalid_argument ("invalid file handle");
#if HAVE_JPEGLIB_H
    init ();
#endif
  }

  jpegstream::~jpegstream (void)
  {
    delete [] _scanline;
#if HAVE_JPEGLIB_H
    if (_header)
      {
        if (0 == _v_sz) lib->finish_compress (&_info);
        lib->destroy_compress (&_info);
      }
#endif
    fflush (_stream);
  }

  basic_imgstream&
  jpegstream::write (const byte_type *line, size_type n)
  {
    if (!line || 0 == n) return *this;
#if HAVE_JPEGLIB_H
    if (!_header)
      {
        write_header ();
      }
    if (!_scanline)
      {
        lib->write_scanlines (&_info, (JSAMPLE **) &line, 1);
        if (0 < _info.err->msg_code)
          throw std::ios_base::failure ("write error");
      }
    else
      {
        // FIXME: assumes that _bits == 1, whereas the condition for
        //        _scanline to be true, see write_init (), requires
        //        only that _bits != 8.
        for (unsigned int i = 0; i < _h_sz; ++i)
          {
            div_t index = div (static_cast<int>(i), static_cast<int>(8 * sizeof (JSAMPLE)));
            int offset = 8 * sizeof (JSAMPLE) - 1 - index.rem;
            _scanline[i] = ((line[index.quot] & (1 << offset))
                            ? 0 : ~0);
          }
        lib->write_scanlines (&_info, (JSAMPLE **) &_scanline, 1);
        if (0 < _info.err->msg_code)
          throw std::ios_base::failure ("write error");
      }
    --_v_sz;
#endif /* HAVE_JPEGLIB_H */
    return *this;
  }

  bool
  jpegstream::is_usable (void)
  {
    if (lib)
      {
        return lib->is_usable;
      }

    lib = new (std::nothrow) jpeg_lib_handle ();
    if (!lib)
      {
        return false;
      }

    lib->is_usable = false;
    lib->message   = string ();
    lib->lib       = NULL;
#if HAVE_JPEGLIB_H
    try
      {
        basic_imgstream::dlopen ("libjpeg", validate);
      }
    catch (std::runtime_error& e)
      {
        lib->message = e.what ();
        return lib->is_usable;
      }
#endif /* HAVE_JPEGLIB_H */

    return lib->is_usable;
  }

#if HAVE_JPEGLIB_H
#define funcsym(name)                                           \
  lib->name                                                     \
    = ((jpegstream::jpeg_lib_handle::name##_f)                  \
       basic_imgstream::dlsym (lib->lib, "jpeg_"#name));
#endif

  bool
  jpegstream::validate (lt_dlhandle h)
  {
    if (!h) return false;

#if HAVE_JPEGLIB_H
    lib->lib = h;

# ifndef jpeg_create_compress
    funcsym (create_compress);
# else
    funcsym (CreateCompress);
# endif
    funcsym (finish_compress);
    funcsym (destroy_compress);
    funcsym (destroy);
    funcsym (stdio_dest);
    funcsym (std_error);
    funcsym (write_scanlines);
    funcsym (set_defaults);
    funcsym (start_compress);
    funcsym (default_qtables);

    // restrict usage of libjpeg to the version range it was compiled against;
    // either before version 7.0 or 7.0 and later
    bool is_version_consistent =
        ((JPEG_LIB_VERSION <  70 && !lib->default_qtables) ||
         (JPEG_LIB_VERSION >= 70 &&  lib->default_qtables));

    lib->is_usable = (
# ifndef jpeg_create_compress
                      lib->create_compress
# else
                      lib->CreateCompress
# endif
                      && lib->finish_compress
                      && lib->destroy_compress
                      && lib->destroy
                      && lib->stdio_dest
                      && lib->std_error
                      && lib->write_scanlines
                      && lib->set_defaults
                      && lib->start_compress
                      && is_version_consistent);
#endif /* HAVE_JPEGLIB_H */

    return lib->is_usable;
  }

#if HAVE_JPEGLIB_H
#undef funcsym
#endif

  basic_imgstream&
  jpegstream::write_header (void)
  {
    check_consistency ();

#if HAVE_JPEGLIB_H

    _info.image_width  = _h_sz;
    _info.image_height = _v_sz;

    _info.in_color_space   = (RGB == _cspc ? JCS_RGB : JCS_GRAYSCALE);
    _info.input_components = (RGB == _cspc ? 3 : 1);

    lib->set_defaults (&_info);

    size_type density_max = (1 << sizeof (_info.X_density) * 8) - 1;
    _info.density_unit = 1;
    _info.X_density = (_hres <= density_max) ? _hres : density_max;
    _info.Y_density = (_vres <= density_max) ? _vres : density_max;

    lib->start_compress (&_info, true);

    if (mono == _cspc && 8 != _bits)
      {
        _scanline = new byte_type[_h_sz];
      }
    else
      {
        _scanline = NULL;
      }
#endif /* HAVE_JPEGLIB_H */

    _header = true;
    return *this;
  }

  void
  jpegstream::check_consistency (void) const
  {
    if (!(mono == _cspc || grey == _cspc || RGB == _cspc))
      {
        throw std::logic_error ("unsupported colour space");
      }
  }

  void
  jpegstream::init (void)
  {
    if (!is_usable ())
      {
        throw std::runtime_error (lib->message);
      }

#if HAVE_JPEGLIB_H
    // set up JPEG library default error handlers first, then override
    // error handling for fatal errors (because the default would just
    // end up calling exit())
    _info.err = lib->std_error (&_err);
    _err.error_exit = error_exit;
# ifndef jpeg_create_compress
    lib->create_compress (&_info);
# else
    lib->CreateCompress (&_info, JPEG_LIB_VERSION,
                         (size_t) sizeof (struct jpeg_compress_struct));
# endif
    lib->stdio_dest (&_info, _stream);
#endif /* HAVE_JPEGLIB_H */
  }

  jpegstream::jpeg_lib_handle *jpegstream::lib = NULL;

#if HAVE_JPEGLIB_H
  void
  jpegstream::error_exit (jpeg_common_struct *info)
  {
    jpegstream::lib->destroy (info);
  }
#endif

} // namespace iscan
